report/4.Solution/GeneratedCppCode.tex

```diff
@@ -0,0 +1,86 @@
+% !TeX root = ../main.tex
+
+\subsection{Generated C++ code}
+While the previous section is primarily focused on the code generation from an abstract point of view, this section focuses on the generated C++ code.
+Since C++ is the output language of the HCL compiler, all the language concepts of HCL have to be represented in valid C++ code.
+
+\textbf{Primitive types}
+The primitive types of HCL includes numbers, text and booleans. 
+Booleans are straightforward, as HCL's boolean behaves exactly like the C++ boolean type.
+This means that the regular C++ boolean type can be used as the output type of HCL booleans.
+
+As HCL aims for simplicity, only one number type is introduced. 
+This type needs to be able to store both integers and non-integers. 
+The simple solution is to turn all HCL numbers into C++'s double type.
+The solution with doubles is not ideal, but it will fulfill the requirements of the number type in HCL.
+
+C++ does have an implementation of a string type in its standard library.
+However, the Arduino does not include this implementation.
+The Arduino uses its own string implementation, which varies from the standard implementation.
+
+As the string type shares a lot of functionality with that of lists, the representation of strings in the generated C++ code, is based on the ConstList.
+
+\textbf{Lists}
+
+For lists, C++ includes a list type in its standard library called vector. 
+This vector type is, like the string type, not available in the libraries provided by the Arduino.
+Furthermore the vector implementation is mutable and the lists in HCL are immutable, a simple implementation of a constant list is implemented instead.
+
+One thing to be aware of when using lists, is that the size of the list cannot always be determined at compile time.
+This means that lists will usually do heap allocation. 
+With heap allocation comes the need to do proper memory handling, eg.
+clean up allocated memory in order to avoid memory leaks.
+
+Memory handling is done by utilizing the C++ concept of RAII\footnote{Resource acquisition is initialization} along with C++ smart pointers.
+This means that a pointer is implemented, which keeps a count of the references to the memory pointed to.
+If the counter goes to zero, this means that there are no more references to memory, and the memory may be deallocated safely.
+
+It should be noted, that the current HCL compiler uses the std::shared\_ptr, which is not available in the Arduino standard library.
+The implementation should however be easily ported to the Arduino, however it is not complete as of the time of writing.
```

Comment:
(Maybe) add a little bit of text explaining how the implementation can be ported.